r/tests/testthat/test-dplyr-arrange.R

```diff
@@ -139,18 +139,14 @@ test_that("arrange() on integer, double, and character columns", {
       collect(),
     tbl
   )
-  expect_warning(
-    expect_equal(
-      tbl %>%
-        Table$create() %>%
-        arrange(abs(int), dbl) %>%
-        collect(),
-      tbl %>%
-        arrange(abs(int), dbl) %>%
-        collect()
-    ),
-    "not supported in Arrow",
-    fixed = TRUE
+  expect_equal(
```

Comment:
If we're going to keep this, could we use `expect_dplyr_equal()` here as well?

Though I'm not certain we really need to have this test since we test that `abs()` works in test-dplyr.R. It might still be good/nice to have a test that confirms the warning when a function is not supported in Arrow which in some ways looks like what this test was really intending to test. Maybe instead of using a function like this that we might make a kernel for we make up some other function that we will never use in a kernel (at least the name) and keep the test that throws a warning + pulls things into R + compares that the results are equal.

Reply:
Looking into `dplyr-arrange.R` I would agree that one test needs to be left as there is a functionality in `arrange()` to check for things not supported by Arrow and this seems to be the only case testing it.

https://github.com/apache/arrow/blob/f8661e032902a963b0a6a46077d72e804d22560d/r/R/dplyr-arrange.R#L39-L46

And so it doesn't make sense to keep `abs()` as it is now supported and it was intended to test for `"try-error"` by `arrow_eval()`. 

https://github.com/apache/arrow/blob/f8661e032902a963b0a6a46077d72e804d22560d/r/R/dplyr-eval.R#L18-L48

We could change `abs()` to for example `ceiling()`. But if it will ever be supported in Arrow this will have to be changed again? And I am not sure how to find a bulletproof function.

Reply:
I was thinking about this myself, and am wondering if that could be achieved via a user-defined function called something ridiculous like `not_a_function` or similar?

Reply:
Thought about this also @thisisnic but then it wouldn't get pull into R or would it?

Reply:
So, there are 2 possible ways that code gets pulled into R when trying to evaluate an `arrow_dplyr_query`:

1. the use of the function results in an error if the input is an Arrow object, and that error is caught https://github.com/apache/arrow/blob/59c5781cbb7c3f806f34b3400d2f4d896c902686/r/R/dplyr-mutate.R#L60-L65

2. other functions that should be pulled into R as they aren't implemented in our NSE functions yet, but calling the base R implementation gives us incorrect results but does not raise an error, are manually specified here: https://github.com/apache/arrow/blob/59c5781cbb7c3f806f34b3400d2f4d896c902686/r/R/dplyr-eval.R#L81-L87

This means, if you create a function that causes an error only when its input is an Arrow object, then that should work for this example.  Hope that makes sense, but if not, I can type up a brief example - just let me know! :)

Reply:
@thisisnic if I understand correctly it should be something in line off:

```R
  dummy_function <- function(x){
    if(assert_is(x, c("numeric", "integer"))) warning()
    x
  }
  expect_warning(
    expect_equal(
      tbl %>%
        Table$create() %>%
        arrange(dummy_function(int), dbl) %>%
        collect(),
      tbl %>%
        arrange(dummy_function(int), dbl) %>%
        collect()
  ),
  "not supported in Arrow",
  fixed = TRUE
  )
```

Does it make sense?

Reply:
Hi @AlenkaF - I think I owe you an apology! 

What I had in mind was a more general way of testing that functions that shouldn't work in the dplyr pipelines, and the code example I sketched out myself uses `mutate` and a function similar to yours there but which results in an error.

However, looking at the tests for `mutate`, this already seems to exist and just uses a function that we don't implement all the arguments for. https://github.com/apache/arrow/blob/f1e3c2e557bb3622b203fb270bf0329f53f1e617/r/tests/testthat/test-dplyr-mutate.R#L121-L137

I don't think this actually can be done in `arrange` as `desc` is already implemented and as far as I'm aware that's the only function you can supply to `arrange` to change the ordering.

Sorry about that, I appreciate you've spent time looking into this!

Reply:
@thisisnic @jonkeane should I put the correction of the `arrange` test - where a solution needs to be found for confirming the warning - in a separate issue and we close this PR?

Reply:
The way `arrange()` works, it should be able to take any valid expressions as arguments, and assuming we have the translations for those functions implemented here in the arrow package, it should work. The `desc()` function is a special case—it is _not_ implemented generally for other dplyr verbs and it only works within `arrange()`.

So I'm confused; is `arrange(abs(int), dbl))` not working? It should work.

Reply:
Thank you for looking into this @ianmcook. `arrange(abs(int), dbl)` in the test works. But it looks as if this part of the test is trying to confirm a warning and in this case `abs()` should be changed?

Reply:
Oh, I understand now 😄 Sorry for my earlier confusion @AlenkaF!

Here is what I would recommend doing:
1. Remove the test code beginning on line 142, and do not replace it with anything. As @jonkeane said in the first comment in this thread:
    > I'm not certain we really need to have this test since we test that `abs()` works in test-dplyr.R

    I agree with that. Actually, I am certain we do not need to have this test.
  
2. Set aside the questions about how best to write tests for unimplemented functions. As @thisisnic says
    >What I had in mind was a more general way of testing functions that shouldn't work in the dplyr pipelines... However, ...  this already seems to exist

    We can revisit that issue separately from this PR, if needed.

Thanks!